### PR TITLE
[None][feat] AutoDeploy: Add Nemotron MOE support for AutoDeploy

### PR DESCRIPTION
**Pull request by nvchenghaoz**
## Release Notes

* **New Features**
  * Added Mixture-of-Experts (MoE) model support with optimized fused kernels and expert routing.
  * Added gated RMS normalization operator for improved normalization efficiency.
  * Enhanced KV cache sizing with memory-aware optimization.
  * Split attention pattern matching into repeat-key-value and standard variants for better optimization control.

* **Tests**
  * Added comprehensive unit tests for MoE kernels, gated normalization, and attention pattern matching.

This is a mass merge to support the Nemotron MOE structure. The key change includes:

1. Support MOE with GEMM - RELU2 - GEMM connection, including updating the torch moe kernel and add the new triton kernel. 
2. Update the rms norm matcher. 
3. Update the attention matcher. 
4. Update the kv cache to support the hybrid model.

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/transform/library/kvcache.py [hunk 1]
```diff
@@ -280,8 +288,16 @@ def _get_mem_info_in_mb():
         memory_for_forward_pass = free_mem_pre - free_mem_post
         self._log_info(f"Memory for forward pass (MB): {memory_for_forward_pass}")
 
-        new_cache_size = free_mem_post * 1024 * 1024 * free_mem_ratio + current_cache_size
-        new_num_pages = int(new_cache_size // (current_cache_size // current_num_pages))
+        # Compute new pages using KV-only bytes to avoid SSM/conv inflating per-page cost
+        # Reserve headroom to avoid OOM from other allocations (workspaces, cudagraph pools, etc.)
+        reserve_mb = max(1024, (total_mem * 5) // 100)  # at least 1 GiB or 5% of total
```
Comment (suyoggupta) [inline tensorrt_llm/_torch/auto_deploy/transform/library/kvcache.py]: Is this strictly needed, or could allocation for cudagraphs, workspaces etc be handled through the existing free_mem_ration flag?

File: tensorrt_llm/_torch/auto_deploy/custom_ops/moe_align_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    if (expert_id <= num_experts)
+    {
+        cumsum[expert_id] = cumsum_val;
+    }
+
+    if (expert_id == num_experts)
+    {
+        *total_tokens_post_pad = cumsum_val;
+    }
+
```
[the remark this answers is not in the document]
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/moe_align_kernel.cu]: Not functional issue in the code... Will skip this suggestion.

File: tensorrt_llm/_torch/auto_deploy/custom_ops/torch_moe.py [hunk 3]
```diff
@@ -1,9 +1,31 @@
...
+        return F.silu
+    elif key == "relu2":
+
+        def relu2(x: torch.Tensor) -> torch.Tensor:
+            return torch.square(F.relu(x))
+
+        return relu2
+    else:
+        raise ValueError(f"Unsupported activation '{name}'. Use one of: silu, relu2.")
+
```
[the remark this answers is not in the document]
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/torch_moe.py]: Good catch

File: tensorrt_llm/_torch/auto_deploy/custom_ops/triton_moe.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
+"""
+Triton implementation of the Fused MOE ops. Inspired by vLLM's triton MOE implementation.
+"""
```
[the remark this answers is not in the document]
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/triton_moe.py]: Same as above, not required?

File: tensorrt_llm/_torch/auto_deploy/transform/library/rms_norm.py [hunk 5]
```diff
@@ -146,3 +147,91 @@ def dummy_args(input_dtype: torch.dtype, weight_dtype: torch.dtype, eps: float =
...
+def _gated_rmsnorm_replacement(
+    x: torch.Tensor,
+    weight: torch.Tensor,
+    gate: torch.Tensor,
+    eps: float,
+    group_size: int,
+) -> torch.Tensor:
+    return torch.ops.auto_deploy.torch_rmsnorm_gated(
+        x, weight, gate, float(eps), int(group_size), False
+    )
```
[the remark this answers is not in the document]
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/transform/library/rms_norm.py]: Not needed.

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/models/test_nemotron_h_patches.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import types
```
[the remark this answers is not in the document]
Comment (nvchenghaoz) [reply tests/unittest/_torch/auto_deploy/unit/singlegpu/models/test_nemotron_h_patches.py]: Same as above..

File: tensorrt_llm/_torch/auto_deploy/transform/library/kvcache.py (hunk 1 above)
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/transform/library/kvcache.py]: It is not strictly needed, but during my debug this would help if the max_num_tokens is set to correct number. I will get back to it later today. 

But as of now, it is fine to keep this as a guard.

**Review by suyoggupta — APPROVED** [review]

**Review by lucaslie — APPROVED** [review]
looks great

**Review by atrifex — APPROVED** [review]
Conditional on no new dependencies being added.

**Comment by nvchenghaoz** [comment]
The .cu file is added for triton moe BF16 support which gives the speed up during the end-to-end model run.
